### PR TITLE
JAMES-2866 Extensions should share singleton

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
Merged

File: server/container/guice/guice-utils/src/main/java/org/apache/james/utils/GuiceGenericLoader.java [hunk 1]
```diff
@@ -98,6 +98,7 @@ public GuiceGenericLoader(Injector injector, ExtendedClassLoader extendedClassLo
         this.additionalExtensionBindings = Modules.combine(extensionConfiguration.getAdditionalGuiceModulesForExtensions()
             .stream()
             .map(Throwing.function(this::<Module>instantiateNoChildModule))
+            .peek(module -> LOGGER.info("Enabling injects contained in " + module.getClass().getCanonicalName()))
```
Comment (mbaechler) [inline server/container/guice/guice-utils/src/main/java/org/apache/james/utils/GuiceGenericLoader.java]: please use placeholder for logs

File: server/container/guice/mailet/src/test/java/org/apache/james/utils/GuiceMailetLoaderTest.java [hunk 2]
```diff
@@ -183,4 +184,25 @@ public void getMailetShouldAllowCustomInjections() throws Exception {
...
+        Mailet mailet1 = guiceMailetLoader.getMailet(FakeMailetConfig.builder()
+            .mailetName("MyGenericMailet")
+            .mailetContext(FakeMailContext.defaultContext())
+            .build());
+        Mailet mailet2 = guiceMailetLoader.getMailet(FakeMailetConfig.builder()
+            .mailetName("MyGenericMailet")
+            .mailetContext(FakeMailContext.defaultContext())
+            .build());
+
+        assertThat(mailet1).isEqualTo(mailet2);
```
Comment (mbaechler) [inline server/container/guice/mailet/src/test/java/org/apache/james/utils/GuiceMailetLoaderTest.java]: you should make the assertion more obvious:
```
assertThat(mailet1.dependency).isSameAs(mailet2.dependency);
```

File: server/container/guice/mailet/src/test/java/org/apache/james/utils/GuiceMailetLoaderTest.java [hunk 3]
```diff
@@ -183,4 +184,25 @@ public void getMailetShouldAllowCustomInjections() throws Exception {
 
         assertThatCode(() -> mailet.service(FakeMail.defaultFakeMail())).doesNotThrowAnyException();
     }
+
+    @Ignore("JAMES-2866 singleton are not shared between extensions")
+    @Test
+    public void allMailetsShouldShareTheSameSingleton() throws Exception {
```
Comment (mbaechler) [inline server/container/guice/mailet/src/test/java/org/apache/james/utils/GuiceMailetLoaderTest.java]: what about using ShrinkWrap to remove the pain of building the extension archive ( http://arquillian.org/arquillian-core/#deployment-archives ) ?

File: server/container/guice/mailet/src/test/java/org/apache/james/utils/GuiceMailetLoaderTest.java (hunk 2 above)
Comment (chibenwa) [reply server/container/guice/mailet/src/test/java/org/apache/james/utils/GuiceMailetLoaderTest.java]: Sure, I thought about it, but then I need to know that it is a MyCustomMailet, and also to have the dependency on the classpath to be able to call non API method.

Here I prefer adding a description.

Comment (mbaechler) [reply server/container/guice/mailet/src/test/java/org/apache/james/utils/GuiceMailetLoaderTest.java]: Ok, I understand, but using equals for that purpose sounds really brittle

**Pull request closed without merging**